polygeist/lib/Dialect/Polygeist/Transforms/SYCLRaiseHost.cpp

```diff
@@ -155,9 +158,78 @@ static bool isClassType(Type type, StringRef className) {
 }
 
 namespace {
-struct RaiseKernelName : public OpRewritePattern<LLVM::AddressOfOp> {
+template <typename SourceOp>
+class OpOrInterfaceRaisePatternBase : public RewritePattern {
+public:
+  using RewritePattern::RewritePattern;
+
+  /// Wrappers around the RewritePattern methods that pass the derived op type.
+  void rewrite(Operation *op, PatternRewriter &rewriter) const final {
+    rewrite(cast<SourceOp>(op), rewriter);
+  }
+  LogicalResult match(Operation *op) const final {
+    // Do not run raising patterns in device code
+    return isInDeviceModule(op) ? failure() : match(cast<SourceOp>(op));
+  }
+  LogicalResult matchAndRewrite(Operation *op,
+                                PatternRewriter &rewriter) const final {
+    // Do not run raising patterns in device code
+    return isInDeviceModule(op) ? failure()
+                                : matchAndRewrite(cast<SourceOp>(op), rewriter);
+  }
+
+  /// Rewrite and Match methods that operate on the SourceOp type. These must be
+  /// overridden by the derived pattern class.
+  virtual void rewrite(SourceOp op, PatternRewriter &rewriter) const {
```

Comment:
Should we use CRTP for these functions to avoid virtual calls?

Reply:
This is a copy from [MLIR's `OpOrInterfaceRewritePatternBase`](https://github.com/intel/llvm/blob/8852c79c3817652e6eec744e083b8ea6dd3a8fa3/mlir/include/mlir/IR/PatternMatch.h#L317). I can still apply your approach if you prefer.

Reply:
Ok, no need to change it in that case 👍